core/node/rpc/sync/client/syncer_set.go

```diff
@@ -0,0 +1,268 @@
+		syncerTasks sync.WaitGroup
+		// muSyncers guards syncers and streamID2Syncer
+		muSyncers sync.Mutex
+		// syncers is the existing set of syncers, indexed by the syncer node address
+		syncers map[common.Address]StreamsSyncer
+		// streamID2Syncer maps from a stream to its syncer
+		streamID2Syncer map[StreamId]StreamsSyncer
+	}
+
+	// SyncCookieSet maps from a stream id to a sync cookie
+	SyncCookieSet map[StreamId]*SyncCookie
+	// StreamCookieSetGroupedByNodeAddress is a mapping from a node address to a SyncCookieSet
+	StreamCookieSetGroupedByNodeAddress map[common.Address]SyncCookieSet
+)
+
+func (cs SyncCookieSet) AsSlice() []*SyncCookie {
+	cookies := make([]*SyncCookie, 0, len(cs))
+	for _, cookie := range cs {
+		cookies = append(cookies, cookie)
+	}
+	return cookies
+}
+
+// NewSyncers creates the required syncer set that subscribe on all given cookies.
+// A syncer can either be local or remote and writes received events to an internal messages channel from which events
+// are streamed to the client.
+func NewSyncers(
+	ctx context.Context,
+	syncID string,
+	streamCache events.StreamCache,
+	nodeRegistry nodes.NodeRegistry,
+	localNodeAddress common.Address,
+	cookies StreamCookieSetGroupedByNodeAddress,
+) (*SyncerSet, <-chan *SyncStreamsResponse, error) {
+	var (
+		syncers         = make(map[common.Address]StreamsSyncer)
+		streamID2Syncer = make(map[StreamId]StreamsSyncer)
+		messages        = make(chan *SyncStreamsResponse, 128)
+	)
+
+	// instantiate background syncers for sync operation
+	for nodeAddress, cookieSet := range cookies {
+		if nodeAddress == localNodeAddress { // stream managed by this node
+			syncer, err := newLocalSyncer(ctx, localNodeAddress, streamCache, cookieSet.AsSlice(), messages)
+			if err != nil {
+				return nil, nil, err
+			}
+			syncers[nodeAddress] = syncer
+		} else {
+			client, err := nodeRegistry.GetStreamServiceClientForAddress(nodeAddress)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			syncer, err := newRemoteSyncer(ctx, syncID, nodeAddress, client, cookieSet.AsSlice(), messages)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			syncers[nodeAddress] = syncer
+		}
+
+		// associate syncer with streamId to remove stream from sync operation
+		syncer := syncers[nodeAddress]
+		for streamID := range cookieSet {
+			streamID2Syncer[streamID] = syncer
+		}
+	}
+
+	return &SyncerSet{
+		ctx:              ctx,
+		syncID:           syncID,
+		streamCache:      streamCache,
+		nodeRegistry:     nodeRegistry,
+		localNodeAddress: localNodeAddress,
+		syncers:          syncers,
+		streamID2Syncer:  streamID2Syncer,
+		messages:         messages,
+	}, messages, nil
+}
+
+func (ss *SyncerSet) Run() {
+	ss.muSyncers.Lock()
+	ss.syncerTasks.Add(len(ss.syncers))
+	for syncerAddr, syncer := range ss.syncers {
+		go func() {
+			syncer.Run()
+			ss.muSyncers.Lock()
+			delete(ss.syncers, syncerAddr)
+			ss.muSyncers.Unlock()
+			ss.syncerTasks.Done()
+		}()
```

Comment:
does this capture work correctly? should `syncerAddr, syncer` be args to the goroutine instead?

Reply:
From go 1.22 this isn't an issue anymore.